File: pset1/credit/credit.c
#include <math.h>
#include <cs50.h>
#include <stdio.h>


int length(long n) ;
int ind_sum(int n) ;
int checksum_alg(int n) ;
int first2digits(long n) ;
int firstdigit(long n) ;



int main()
{
  //*****************//
    long digit ;
    int i, j ;
    int digit2 ;
    long card ;
    int sum1 = 0 ;
    int sum2 = 0 ;
    int sumeven ;
    int sumodd = 0;
    int digit0 ;
    int total ;



    //***************************// to get input from user
    do
    {
        card = get_long("Credit card number : ");
    }
    while (card < 0);  //pow(10,12)




//********************************************// the checksum algorithm
   for(i = 0 ; i < length(card) -1 ; i+=2)
   {
       digit = (card % (long)pow(10,i+2) - card % (long)pow(10,i+1))/ (long)pow(10,i+1);
       digit2 = digit * 2 ;
       if(digit2 <= 9){
           sum1 = sum1 + digit2 ;
       }
       else
       {
           sum2 = sum2 + ind_sum(digit2) ;
       }

   }
   sumeven = sum1 + sum2 ;
   //printf("\n %i \n ", sumeven ) ;

  //Sum of odd place digit

   for(j = 0 ; j < length(card) ; j+=2)
   {
       digit0 = (card % (long)pow(10,j+1) - card % (long)pow(10,j))/ (long)pow(10,j);
       sumodd = sumodd + digit0 ;
   }
   //printf("\n %i \n ", sumodd ) ;

    total = sumodd + sumeven ;
    //printf("\n %i \n ",total);

   //**************************************************// check if modulo 10 = zero
   //printf("\n %i \n ",length(card));
   if(total % 10 !=0)
   {
       printf("INVALID\n");
   }
   else
   {
        if(length(card) != 13 && length(card) != 15 && length(card) != 16 )
        {
             printf("INVALID\n") ;

        }
        else
        {
            if(length(card) == 15 && (first2digits(card) == 34 || first2digits(card)== 37 ))
            {
                printf("AMEX\n");
            }
            else if (length(card) == 16 && (first2digits(card) == 51 || first2digits(card) == 52 || first2digits(card) == 53 || first2digits(card) == 54 || first2digits(card) == 55 ))
            {
                printf("MASTERCARD\n") ;
            }
            else if((length(card) == 13 || length(card) == 16) && firstdigit(card) == 4)
            {
                printf("VISA\n") ;
            }
            else
            {
                printf("INVALID\n") ;
            }
        }
   }




}



//MY FUNCTIONS
int length(long n)
{
    int i = 0 ;
    while(n != 0)
    {
        n = n/10 ;
        i++ ;
    }
    return i ;
}

int ind_sum(int n)
{
    int sum = 0 ;
    while(n != 0)
    {
        sum = sum + n % 10 ;
        n = n/10 ;
    }
    return sum ;
}

int first2digits(long n)
{
    long i = n ;
    while (i >= 100)
    {
        i /= 10;
    }

    return i;
}

int firstdigit(long n)
{
    long i = n ;
    while (i >= 10)
    {
        i /= 10;
    }

    return i;
}

File: pset6/Credit/credit.py
from cs50 import get_int

def main():
  number = get_int("Enter the number: ")
  cardType(number)

def is_valid(number):
  Num = []
  for digit in str(number):
    Num.append(int(digit))
 #PART 1 
  RevNum = Num[::-1]
  temp = RevNum[1::2]
  EvenList = [i*2 for i in temp]
  

  #SINGLE DIGIT LIST
  for element in range(len(EvenList)):
    if EvenList[element] > 9 :
      sum = 0
      for digit in str(EvenList[element]):
          sum = sum + int(digit)
      EvenList[element] = sum
      

  #SUM OF SINGLE DIGIT LIST
  sumEven = 0
  for digit in EvenList:
    sumEven = sumEven +  digit

 #PART 2
  OddList = RevNum[::2] 
  sumOdd = 0
  for digit in OddList:
    sumOdd = sumOdd + digit
  
 #PART 3
  SUM = sumEven + sumOdd

  if SUM % 10 == 0:
    return True

def cardType(number):
  if is_valid(number) == True:
    Num = []
    for digit in str(number):
      Num.append(int(digit))
    if (len(Num) == 15 and Num[0] == 3 and (Num[1] == 4 or Num[1] == 7)):
      print("AMEX")
    elif (len(Num) == 16 and Num[0] == 5 and (Num[1] == 1 or Num[1] == 2 or Num[1] == 3 or Num[1] == 4 or Num[1] == 5 )):
      print("MASTERCARD")
    elif ((len(Num) == 13 or len(Num) == 16) and Num[0] == 4):
      print("VISA")
  else :
    print("INVALID")


if __name__ == "__main__":
  main()